clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -637,6 +638,9 @@ private void handleResponse(RequestHeader header, Struct body, long now) {
                 log.trace("Ignoring empty metadata response with correlation id {}.", header.correlationId());
                 this.metadata.failedUpdate(now);
             }
+            for (String topic : response.errorTopics(Errors.UNKNOWN_TOPIC_OR_PARTITION))
+                this.metadata.maybeSetExpiry(topic, now);
```

Comment:
This confused me for awhile because setting the expiration to `now` seemed odd. Thinking about it more, is it possible there are really only 3 states and we don't need to track any timestamps? I think they would be:

* Active and needs a metadata refresh (What is currently called `TOPIC_EXPIRY_NEEDS_UPDATE`)
* Never remove from topic set (Currently `TOPIC_EXPIRY_NONE`)
* Failed one lookup and is now candidate for removal (current setting timestamp to `now`).

And I suppose there is another implicit state -- removed or missing. So I think we could simplify to a simple enum, which might also help explain the state transitions better.